prereise/gather/hiflddata/data_trans.py

```diff
@@ -0,0 +1,645 @@
+    # list1.sort(reverse=True)
+    return Max_nodeSet
+
+
+def InitKV(clean_data):
+    """Calculate the base_KV for each node based on MIN_VOLT and MAX_VOLT;
+    Save the invalid substations as to be calculated later
+
+    :param pandas.DataFrame clean_data: substation dataframe as returned by :func:`Clean`
+    :return: (*dict*) -- KV_dict, a dict of substation's baseKV value.
+    :return: (*list*) -- to_cal, a list of substations to be calculated using neighbour search.
+    """
+
+    KV_dict = {}
+    to_cal = []
+    Max_Value = 3000
+    Min_Value = 0
+    for index, row in clean_data.iterrows():
+        base_KV = default_base_kv
+        Min_Vol = row["MIN_VOLT"]
+        Max_Vol = row["MAX_VOLT"]
+        if Min_Vol >= Max_Value or Min_Vol <= Min_Value:
+            if Max_Vol < Max_Value and Max_Vol > Min_Value:
+                base_KV = Max_Vol
+            else:
+                to_cal.append((row["LATITUDE"], row["LONGITUDE"]))
+                continue
+        else:
+            if Max_Vol < Max_Value and Max_Vol > Min_Value:
+                base_KV = (Max_Vol + Min_Vol) / 2
+            else:
+                base_KV = Min_Vol
+        KV_dict[(row["LATITUDE"], row["LONGITUDE"])] = base_KV
+    return KV_dict, to_cal
+
+
+# Give a node, find its neighbors in 5 iteration
+
+
+def get_neigbors(g, node, depth=1):
+    """Give a node, find its neighbors in given number depth
+
+    :param networkx.Graph g: Graph generated by :func:`GraphOfNet`
+    :param tuple node: substation coordinate tuple
+    :param int depth: number of hops that we are searching the neighbor substations
+    :return: (*list*) --  a list of closest substations.
+    """
+    output = {}
+    layers = dict(nx.bfs_successors(g, source=node, depth_limit=depth))
+    nodes = [node]
+    for i in range(1, depth + 1):
+        output[i] = []
+        for x in nodes:
+            output[i].extend(layers.get(x, []))
+        nodes = output[i]
+    return output
+
+
+def Cal_KV(N_dict, G, KV_dict, to_cal):
+    """Estimate missing substation's KV data using the neighbors average calculation
+
+    :param dict N_dict: dict of substation's neighbors as returned by :func:`Neighbors`
+    :param networkx.Graph g: Graph generated by :func:`GraphOfNet`
+    :param dict KV_dict: dict of existing baseKV for each substation
+    :param list to_cal: list of the substations to be calculate using average KV
+    :return: (*dict*) --  dict of all substation's baseKV being corrected and calculated.
+    """
+
+    for sub in to_cal:
+        if sub not in N_dict:
+            continue
+        neigh = get_neigbors(G, sub, depth=5)
+        temp_KV = 0
+        num = 0
+        for i in range(1, 6):
+            for nei in neigh[i]:
+                if nei in KV_dict:
+                    temp_KV = temp_KV + KV_dict[nei]
+                    num = num + 1
+            if num > 0:
+                KV_dict[sub] = temp_KV / num
+                continue
```

Comment:
Is this equivalent to:
```python
matching_neighbors = set(neigh[i]) & set(KV_dict.keys())
if len(matching_neighbors) > 0:
    KV_dict[sub] = sum(KV_dict[n] for n in matching_neighbors) / len(matching_neighbors)
    continue
```
?

Reply:
Another question: why do we have a `continue` in this for loop, rather than a `break`? I would think once we've found neighbors with good voltages at one depth, we don't need to continue to find neighbors at further depths, right?

Reply:
We should use a `break` for this case. The algo is supposed to do a BFS to approximate the missing base_KV and break early. thanks for finding this.

Reply:
> Is this equivalent to:
> 
> ```python
> matching_neighbors = set(neigh[i]) & set(KV_dict.keys())
> if len(matching_neighbors) > 0:
>     KV_dict[sub] = sum(KV_dict[n] for n in matching_neighbors) / len(matching_neighbors)
>     continue
> ```
> 
> ?

The code is more concise but not quite efficient. I have tried it and it would take long time to process the data. So I prefer to keep it as it is.

Reply:
What is the time difference, and where does it come from? The two seem to be doing the exact same calculations, with the exception of re-generating `set(KV_dict.keys())`, which could probably be moved outside of the loops. This would probably also address the use of voltages that are set in this loop in subsequent loop iterations, by freezing the set of 'real' data before we start filling in estimated data.

Reply:
The time difference is from `set(neigh[i]) & set(KV_dict.keys())`, your are right, no need to perform set operation for `KV_dict.keys()` in the for loop each time. 
Will replace by 
```python
            count = sum(1 for neighbor in neighbors[depth] if neighbor in kv_dict)
            if count > 0:
                sum_kv = sum(kv_dict[nei] for nei in neighbors[depth] if nei in kv_dict)
                kv_dict[sub] = sum_kv / count
                break
```

Reply:
I have explained why it gets slow down and use an improved count for that purpose. Now the sum is applied to simplify these lines with the count. Please raise if any more concern. Resolving this now.